File: src/listener.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "std_msgs/Float32.h"

#include "geometry_msgs/Twist.h"
#include "geometry_msgs/Vector3.h"
#include "sensor_msgs/NavSatFix.h"
#include "geometry_msgs/Pose.h"
#include "control_law.cpp"
#include "calculate_speed.cpp"

control_jellyfishbot jellyfishbot_control_system;
calculate_speads calculate_speads_algorithm;
struct Quaternion {
    double w, x, y, z;
    };

    struct EulerAngles {
        double roll, pitch, yaw;
    };


int following_point = 1;
int counter = 0;
int point1 = 0;
double tau_L = 0;
double tau_R = 0;
double tau_M = 0;



EulerAngles ToEulerAngles(Quaternion q) {
        EulerAngles angles;

        // roll (x-axis rotation)
        double sinr_cosp = 2 * (q.w * q.x + q.y * q.z);
        double cosr_cosp = 1 - 2 * (q.x * q.x + q.y * q.y);
        angles.roll = std::atan2(sinr_cosp, cosr_cosp);

        // pitch (y-axis rotation)
        double sinp = 2 * (q.w * q.y - q.z * q.x);
        if (std::abs(sinp) >= 1)
            angles.pitch = std::copysign(M_PI / 2, sinp); // use 90 degrees if out of range
        else
            angles.pitch = std::asin(sinp);

        // yaw (z-axis rotation)
        double siny_cosp = 2 * (q.w * q.z + q.x * q.y);
        double cosy_cosp = 1 - 2 * (q.y * q.y + q.z * q.z);
        angles.yaw = std::atan2(siny_cosp, cosy_cosp);

        return angles;
    }
int hv = 0;

void velocity_callback(const geometry_msgs::Twist& msg)
{   
    hv = hv +1;
    //ROS_INFO("[Listener] I heard speed: [%f]\n",u);
    jellyfishbot_control_system.u = msg.linear.x;
    jellyfishbot_control_system.v = msg.linear.y;
    jellyfishbot_control_system.r = msg.angular.z;
    //cout << "I got in the velocity callback: "<< hv << endl; 
    
}
int counter_gps = 0;
// void gps_callback(const sensor_msgs::NavSatFix& msg)
//     {   
//     calculate_speads_algorithm.latitude1 = calculate_speads_algorithm.latitude2;
//     calculate_speads_algorithm.longitude1 = calculate_speads_algorithm.longitude2;
//     calculate_speads_algorithm.timeStamp0 = calculate_speads_algorithm.timeStamp;
//     calculate_speads_algorithm.latitude2 = msg.latitude;
//     calculate_speads_algorithm.longitude2 = msg.longitude;
//     calculate_speads_algorithm.timeStamp = msg.header.stamp.sec;
//     cout << "msg.header.stamp " << calculate_speads_algorithm.timeStamp << endl;


//     cout << "heheheheheh" << endl;
//     cout << "lat1: "<< calculate_speads_algorithm.latitude1<< " ,lon1: "<< calculate_speads_algorithm.longitude1 << endl;
//     cout << "lat2: "<< calculate_speads_algorithm.latitude2<< " ,lon2: "<< calculate_speads_algorithm.longitude2 << endl;
//     double distance = calculate_speads_algorithm.CoordinatesToMeters();
//     double velocity = calculate_speads_algorithm.calculate_velocity();
//     double angle = calculate_speads_algorithm.CoordinatesToAngle();
//     cout << "distance: " << distance << endl;
//     cout << "velocity: " << velocity << endl;
//     cout << "angle: " << angle << endl;
    
//     }
int hp = 0;
int xf_10;
int xf_0;
bool change_point = false;
void pose_callback(const geometry_msgs::Pose& msg)
{    
    jellyfishbot_control_system.dt = 0.1;

    hp = hp +1;
    double threshold = 0.5;

    double xq =  msg.orientation.x;
    double yq =  msg.orientation.y;
    double zq =  msg.orientation.z;
    double wq =  msg.orientation.w;
    struct Quaternion q;
    struct EulerAngles EulerAngles;
    q.x = xq;
    q.y = yq;
    q.z = zq;
    q.w = wq;
    EulerAngles = ToEulerAngles(q);
    jellyfishbot_control_system.psi = EulerAngles.yaw;
    jellyfishbot_control_system.x = msg.position.x;
    jellyfishbot_control_system.y = msg.position.y;
    //cout << "I got in the pose callback: "<< hp << endl; 
    

    // ROS_INFO("[Listener] I heard poseX: [%f]\n",x);
    // ROS_INFO("[Listener] I heard poseYaw: [%f]\n",yaw);
    //cout << "I got here: "<< h << endl; 
        double ax, ay, xf, yf;
        double delta = 3;
        if (following_point == 1){
            ax = 0;
            ay = 0;
            yf = 10;
            xf = 0;
            // xf_0 = 0;
            // xf_10 = 0;
            // change_point = false;
        }
        
        // if (following_point > 1){
            
        //     if (following_point%2 == 0){
        //         ax = xf;
        //         ay = yf;
        //         yf = yf;
        //         change_point = false;
        //     }
        //     else{
        //         ax = xf;
        //         ay = yf;
        //         yf = yf + 0.7;
        //         change_point = false;
        //     }

        //     if (following_point = 2){
        //         xf = 10;
        //         xf_10 = 1;
        //         xf_0 = 0;
        //         change_point = false;
        //     }
        //     else if (following_point > 2 && change_point == true){
        //         if (xf_10 == 1 && xf_0 ==0 ){
        //             xf = 10;
        //             xf_10 = 2;
        //             xf_0 = 0;
        //             change_point = false;
        //         } 
        //         else if (xf_10 =2 && xf_0 ==0 ){
        //             xf = 0;
        //             xf_10 = 0;
        //             xf_0 =1;
        //             change_point = false;
        //         }
        //         else if (xf_10 ==0 and xf_0 ==1){
        //             xf = 0;
        //             xf_0 = 2;
        //             xf_10 = 0;
        //             change_point = false;
        //         }
        //         else if (xf_10 == 0 && xf_0 == 2){
        //             xf = 10;
        //             xf_0 = 0;
        //             xf_10 = 1;
        //             change_point = false;
        //         }
        //     }

        // }
        
        else if (following_point == 2){
            ax = 0;
            ay = 10;
            yf = 10;
            xf = 0.7 ;
        }
        else if (following_point == 3){
            ax = 0.7;
            ay = 10;
            yf = 0;
            xf = 0.7 ;
        }
        else if (following_point == 4){
            ax = 0.7;
            ay = 0;
            yf = 0;
            xf = 1.4 ;
        }
        else if (following_point == 5){
            ax = 1.4;
            ay = 0;
            yf = 10;
            xf = 1.4;
        }  
        else if (following_point == 6){
            ax = 1.4;
            ay = 10;
            yf = 10;
            xf = 2.1;
        } 
        else if (following_point == 7){
            ax = 2.1;
            ay = 10;
            yf = 0;
            xf = 2.1;
        } 
        cout << "xf,yf: "<< xf <<","<< yf << endl;
        jellyfishbot_control_system.ax = ax;
        jellyfishbot_control_system.ay = ay;
        jellyfishbot_control_system.yf = yf;
        jellyfishbot_control_system.xf = xf ;
        jellyfishbot_control_system.delta = delta;
        jellyfishbot_control_system.projected_point();
        double xp = jellyfishbot_control_system.xp;
        double yp = jellyfishbot_control_system.yp;
        //jellyfishbot_control_system.virtual_target(ax,ay,xp,yp,xf,yf,delta,1000000,30); //1000000,30
        jellyfishbot_control_system.virtual_target2(ax,ay,xp,yp,xf,yf,delta,1000000,30); //1000000,30

        double xd = jellyfishbot_control_system.xd;
        double yd = jellyfishbot_control_system.yd;
        if (jellyfishbot_control_system.enable_print){
            //cout <<"ax,ay = " << jellyfishbot_control_system.ax <<","<<jellyfishbot_control_system.ay<<endl;
            //cout << "xp,yp = " << xp<< ","<<yp << endl;
            //cout << "xd,yd = " << xd << ","<< yd<< endl;
            //cout << "xf,yf = "<< xf << "," << yf << endl;
        }
        counter = counter +1;
        
        double u_d= 0.5;
        if (jellyfishbot_control_system.e_psi > 0.1 | jellyfishbot_control_system.e_psi<-0.1){
            u_d = 0.07;
        }
            
        cout << "distance to next point: "<< jellyfishbot_control_system.distance << endl;
        // if (jellyfishbot_control_system.distance < 0.5  & counter > 100 | jellyfishbot_control_system.getting_away){
        //     point1 = point1+1;
        //     counter = 0;
        // }

        cout << "Counter: " << counter << endl;
        if (jellyfishbot_control_system.distance < threshold & following_point ==1 & counter > 100 |  jellyfishbot_control_system.getting_away){
            following_point = 2;
            counter = 0;
            change_point = true;
        }
        else if (jellyfishbot_control_system.distance < threshold & following_point == 2 & counter > 100 |  jellyfishbot_control_system.getting_away){
            following_point = 3;
            counter = 0;
            change_point = true;
        }
        else if (jellyfishbot_control_system.distance < threshold & following_point == 3 & counter > 100 |  jellyfishbot_control_system.getting_away){
            following_point = 4;
            counter = 0;
            change_point = true;
        }
        else if (jellyfishbot_control_system.distance < threshold & following_point == 4 & counter > 100 |  jellyfishbot_control_system.getting_away){
            following_point = 5;
            counter = 0;
            change_point = true;
        }
        else if (jellyfishbot_control_system.distance < threshold & following_point == 5 & counter > 100 |  jellyfishbot_control_system.getting_away){
            following_point = 6;
            counter = 0;
            change_point = true;
        }
        else if (jellyfishbot_control_system.distance < threshold & following_point == 6 & counter > 100 |  jellyfishbot_control_system.getting_away){
            following_point = 7;
            counter = 0;
            change_point = true;
        }
        double v_d= 0;
        jellyfishbot_control_system.u_d_yaw = 5; // desired surge velocity for periods of yaw motion


        
        jellyfishbot_control_system.xd = xd ;
        jellyfishbot_control_system.yd = yd;
        jellyfishbot_control_system.u_d = u_d;
        jellyfishbot_control_system.v_d = v_d;
        cout << "u_d: " <<jellyfishbot_control_system.u_d << ",u: " <<jellyfishbot_control_system.u << endl;
        
        if (! jellyfishbot_control_system.check_arrival(threshold) | 1 <2){
            //cout << "P4 " << endl;
            //jellyfishbot_control_system.obtain__thruster_commands_LOS_Virtual_target(u_d,v_d,jellyfishbot_control_system.xd,jellyfishbot_control_system.yd);
            jellyfishbot_control_system.obtain__thruster_commands_LOS_Virtual_target_NO_SPEEDS(u_d,v_d,jellyfishbot_control_system.xd,jellyfishbot_control_system.yd);

            tau_L = jellyfishbot_control_system.tau_L;
            tau_R = jellyfishbot_control_system.tau_R;
            tau_M = jellyfishbot_control_system.tau_M;
            
            if (jellyfishbot_control_system.distance < 0.5 & following_point == 5 & counter > 100 | jellyfishbot_control_system.getting_away){
                jellyfishbot_control_system.stop = true;
                tau_L = 0;
                tau_R = 0;
                tau_M = 0;
                //cout <<"heeey I reached the final point" <<endl;
            }
           
            // pub_thrust_l.publish(msg_tau_R);     
            // pub_thrust_r.publish(msg_tau_L);     
            // pub_thrust_t.publish(msg_tau_M);
            // projected_point_msg = Vector3()
            // projected_point_msg.x = jellyfishbot_control_system.xp
            // projected_point_msg.y = jellyfishbot_control_system.yp
            // projected_point_topic.publish(projected_point_msg)
            // if jellyfishbot_control_system.enable_print:
            //     print("thrusters: {},{},{},    {}, {},{},   {}  {}    {}".format(tau_L,tau_R,tau_M,round(jellyfishbot_control_system.psi_d,3),round(jellyfishbot_control_system.psi,3),round(jellyfishbot_control_system.e_psi,3),jellyfishbot_control_system.movement,round(jellyfishbot_control_system.distance,2),round(jellyfishbot_control_system.u,2)))
           
        }
            std_msgs::Float32 msg_tau_R;
            std_msgs::Float32 msg_tau_L;
            std_msgs::Float32 msg_tau_M;
            geometry_msgs::Vector3 pp_msg;
            geometry_msgs::Vector3 vt_msg;
            msg_tau_L.data = tau_L;
            msg_tau_R.data = tau_R;
            msg_tau_M.data = tau_M;
            pp_msg.x = jellyfishbot_control_system.xp;
            pp_msg.y = jellyfishbot_control_system.yp;

            vt_msg.x = jellyfishbot_control_system.xd;
            vt_msg.y = jellyfishbot_control_system.yd;
            cout << "ZZZ xd,yd: " << jellyfishbot_control_system.xd <<","<<jellyfishbot_control_system.yd<<endl; 

            // msg_tau_L.data = 5;
            // msg_tau_R.data = 5;
            // msg_tau_M.data = 0;
    

        

            jellyfishbot_control_system.pub_thrust_l.publish(msg_tau_R);
            jellyfishbot_control_system.pub_thrust_r.publish(msg_tau_L);
            jellyfishbot_control_system.pub_thrust_tt.publish(msg_tau_M);
            jellyfishbot_control_system.pp_topic.publish(pp_msg);
            jellyfishbot_control_system.vt_topic.publish(vt_msg);
	    
       

       


}
    void listener(ros::NodeHandle node,ros::Publisher pub_thrust_l,ros::Publisher pub_thrust_r,ros::Publisher pub_thrust_tt){
        ros::Subscriber sub1 = node.subscribe("/robot_twist_bff", 1000, velocity_callback);
        ros::Subscriber sub2 = node.subscribe("/robot_pose", 1000, pose_callback);
        //ros::Subscriber sub3 = node.subscribe("/positionGPS", 1000, gps_callback);
        //ros::Subscriber sub3 = node.subscribe("/robot_twist_bff", 1000, velocity_callback2);

        ros::spin();
    }


int main(int argc, char **argv)
{
    // Initiate a new ROS node named "listener"
	ros::init(argc, argv, "listener_node");
	//create a node handle: it is reference assigned to a new node
	ros::NodeHandle node;

    string thL = "/thrust_l";
    string thR = "/thrust_r";
    string thT = "/thrust_t";
    // thL = "/thrust_l0";
    // thR = "/thrust_r0";
    // thT = "/thrust_t0";

    ros::Publisher pub_thrust_l = node.advertise<std_msgs::Float32>(thL, 1000);
    ros::Publisher pub_thrust_r = node.advertise<std_msgs::Float32>(thR, 1000);
    ros::Publisher pub_thrust_tt = node.advertise<std_msgs::Float32>(thT, 1000);
    ros::Publisher pp_topic = node.advertise<geometry_msgs::Vector3>("/projected_point_topic", 1000);
    ros::Publisher vt_topic = node.advertise<geometry_msgs::Vector3>("/virtual_target_topic", 1000);
    
    jellyfishbot_control_system.pub_thrust_l = pub_thrust_l;
    jellyfishbot_control_system.pub_thrust_r = pub_thrust_r;
    jellyfishbot_control_system.pub_thrust_tt = pub_thrust_tt;
    jellyfishbot_control_system.pp_topic = pp_topic;
    jellyfishbot_control_system.vt_topic = vt_topic;
    
    
    // ros::Publisher pub_thrust_l = node.advertise<std_msgs::Float32>(thL, 1000);
    // ros::Publisher pub_thrust_r = node.advertise<std_msgs::Float32>(thR, 1000);
    // ros::Publisher pub_thrust_t = node.advertise<std_msgs::Float32>(thT, 1000);

    // Subscribe to a given topic, in this case "chatter".
	//chatterCallback: is the name of the callback function that will be executed each time a message is received.


         //loop_rate.sleep();
        
       
        
        //loop_rate.sleep(); 

//     while (ros::ok()) 
     while (node.ok()){
     listener(node,pub_thrust_l,pub_thrust_r,pub_thrust_tt);
     }    
            
        
        
//    }

    return 0;
}

File: scripts/listener.py
#!/usr/bin/env python
from calendar import c

from psutil import NoSuchProcess
import rospy
from geometry_msgs.msg import Twist
from geometry_msgs.msg import Pose
from rosgraph_msgs.msg import Clock
from std_msgs.msg import Float32
from geometry_msgs.msg import Vector3
import numpy as np
from tf.transformations import euler_from_quaternion
import control_law
import path_generation
from math import pi, tan, sqrt
from sympy import symbols, Eq, solve
import cmath
from scipy.optimize import fsolve

jellyfishbot_control_system=control_law.control_jellyfishbot()
path = path_generation.path_generation()
path.W = 25
path.L = 10
path.R = 7
path.x0 =6
path.y0 = 6
path.alpha = 0



path.x0 = path.x0/2
path.y0 = path.y0/2
path.enable_overlap()
global path_points
global point
global Npoints

point = 0
path_points = path.generate_points()
Npoints = path.Npoints 
print("Number of points is {}".format(Npoints))
import time
global first_time
global counter
counter = 0
first_time = False
global first_time2
first_time2 = False
global following_point
global point1
point1 = 0
following_point = 1

def velocity_callback(message):
    #rospy.loginfo(rospy.get_caller_id() + " u,v,r are %s",[ message.linear.x,message.linear.y,message.angular.z])
    u = message.linear.x
    v = message.linear.y
    r = message.angular.z
    jellyfishbot_control_system.log_current_velocity(u,v,r)
    
command_thrusters = 1

jellyfishbot_control_system.enable_print = True
def pose_callback(message):
    #rospy.loginfo(rospy.get_caller_id() + " u,v,r are %s",[ message.linear.x,message.linear.y,message.angular.z])
    x = message.position.x
    y = message.position.y
    orientation_q=message.orientation
    orientation_list = [orientation_q.x, orientation_q.y, orientation_q.z, orientation_q.w]
    (roll, pitch, yaw) = euler_from_quaternion (orientation_list)
    jellyfishbot_control_system.log_current_heading(yaw)
    jellyfishbot_control_system.log_current_location(x,y)
    #print(path_points)
    global point
    global first_time
    global first_time2
    global path_points

    MODE = 3

    global counter
    if not first_time and MODE == 3:
        path_points[0,0] = x
        path_points[1,0] = y
        first_time = True
        
        #path_points=np.array([[0 ,10],[0, 0]])
        if jellyfishbot_control_system.enable_print:
            print(path_points)
        path.plot_path()
    

    
    ############ MODE 1 - One point and then stops ############
    if MODE == 1: 
        
        global first_time2
        xd = 0
        yd = 0
        jellyfishbot_control_system.xf = xd
        jellyfishbot_control_system.yf = yd
        
        u_d= 0.5
        v_d= 0
        jellyfishbot_control_system.u_d_yaw = 5 # desired surge velocity for periods of yaw motion


        threshold = 0.5
        
        jellyfishbot_control_system.xd = xd 
        jellyfishbot_control_system.yd = yd
        jellyfishbot_control_system.u_d = u_d
        jellyfishbot_control_system.v_d = v_d
        if not jellyfishbot_control_system.check_arrival(threshold):# or 1 <2:

            tau_L,tau_R,tau_M = jellyfishbot_control_system.obtain__thruster_commands_LOS_Virtual_target()

            pub_thrust_l.publish(tau_R)     
            pub_thrust_r.publish(tau_L)     
            pub_thrust_t.publish(tau_M)
            if jellyfishbot_control_system.enable_print:
                print("thrusters: {},{},{},    {}, {},{},   {}  {}      {}".format(tau_L,tau_R,tau_M,round(jellyfishbot_control_system.psi_d,3),round(jellyfishbot_control_system.psi,3),round(jellyfishbot_control_system.e_psi,3),jellyfishbot_control_system.movement,round(jellyfishbot_control_system.distance,2),round(jellyfishbot_control_system.u,2)))
        else:
            
            jellyfishbot_control_system.stop = True
            tau_L,tau_R,tau_M = 0,0,0
            if jellyfishbot_control_system.enable_print:
                print("I reach the point {},{}, haaay".format(xd,yd))
   


    elif MODE == 2:
        #print('555')
        # waypoints
        global first_time2
        global following_point

                

        ####
        delta = 1
        #if  first_time2:
        if following_point == 1:
            ax = 0
            ay = 0
            yf = 0
            xf = -10 
        if following_point == 2:
            ax = -10
            ay = 0
            yf = -10
            xf = -10 
        elif following_point == 3:
            ax = -10
            ay = -10
            yf = -10
            xf = 10 
        elif following_point == 4:
            ax = 10
            ay = -10
            yf = 10
            xf = 10  
        elif following_point == 5:
            ax = 10
            ay = 10
            yf = 0
            xf = 0  
        #jellyfishbot_control_system.Yp = Yp 
        jellyfishbot_control_system.ax = ax
        jellyfishbot_control_system.ay = ay
        jellyfishbot_control_system.yf = yf
        jellyfishbot_control_system.xf = xf 
        jellyfishbot_control_system.delta = delta

        
            

        projected_point = jellyfishbot_control_system.projected_point3()

        yp = projected_point[0] #+ ay
        xp = projected_point[1]


        jellyfishbot_control_system.xp = xp
        jellyfishbot_control_system.yp = yp

        yd,xd = jellyfishbot_control_system.virtual_target2()
     
     
        if jellyfishbot_control_system.enable_print:
            print("ax,ay = {},{}".format(jellyfishbot_control_system.ax,jellyfishbot_control_system.ay))
            print("xp,yp = {},{}".format(xp,yp))
            print("xd,yd = {},{}".format(xd,yd))
            print("xf,yf = {},{}".format(jellyfishbot_control_system.xf,jellyfishbot_control_system.yf))
        global counter
            
        counter = counter +1
        
        u_d= 0.5
        cross_track_error = sqrt((xp - x)**2+(yp - y)**2)
        #print("Cross track error: {}".format(cross_track_error))
        if jellyfishbot_control_system.distance < 1 or cross_track_error > 0.5 or counter < 100 :
            u_d = 0.1
        elif 100 <= counter < 150:
            u_d = 0.3
        #print("counter: {}".format(counter))
  
  
        if jellyfishbot_control_system.distance < 0.5 and following_point ==1 and counter > 100 or jellyfishbot_control_system.getting_away:
            following_point = 2
            counter = 0
        elif jellyfishbot_control_system.distance < 0.5 and following_point == 2 and counter > 100 or jellyfishbot_control_system.getting_away:
            following_point = 3
            counter = 0
        elif jellyfishbot_control_system.distance < 0.5 and following_point == 3 and counter > 100 or jellyfishbot_control_system.getting_away:
            following_point = 4
            counter = 0
        elif jellyfishbot_control_system.distance < 0.5 and following_point == 4 and counter > 100 or jellyfishbot_control_system.getting_away:
            following_point = 5
            counter = 0
        v_d= 0
        jellyfishbot_control_system.u_d_yaw = 5 # desired surge velocity for periods of yaw motion


        threshold = 0.5
        jellyfishbot_control_system.xd = xd 
        jellyfishbot_control_system.yd = yd
        jellyfishbot_control_system.u_d = u_d
        jellyfishbot_control_system.v_d = v_d
        if not jellyfishbot_control_system.check_arrival(threshold) or 1 <2:
            
            #tau_L,tau_R,tau_M = jellyfishbot_control_system.obtain__thruster_commands_LOS_V4(xd,yd,u_d,v_d)
            tau_L,tau_R,tau_M = jellyfishbot_control_system.obtain__thruster_commands_LOS_Virtual_target(u_d)
            
            #tau_R = min(tau_R,35)
            #tau_L = min(tau_L,35)
            
            pub_thrust_l.publish(tau_R)     
            pub_thrust_r.publish(tau_L)     
            pub_thrust_t.publish(tau_M)
            if jellyfishbot_control_system.enable_print:
                print("thrusters: {},{},{},    {}, {},{},   {}  {}    {}".format(tau_L,tau_R,tau_M,round(jellyfishbot_control_system.psi_d,3),round(jellyfishbot_control_system.psi,3),round(jellyfishbot_control_system.e_psi,3),jellyfishbot_control_system.movement,round(jellyfishbot_control_system.distance,2),round(jellyfishbot_control_system.u,2)))
                
        else:
            following_point = following_point +1
            #jellyfishbot_control_system.stop = True
            #tau_L,tau_R,tau_M = 0,0,0
            if jellyfishbot_control_system.enable_print:
                 print("I reach the point {},{}, haaay".format(jellyfishbot_control_system.xf,jellyfishbot_control_system.yf))
 
        # jellyfishbot_control_system.distance0 = distance0
        # print("distance0,distance: {},{}".format(jellyfishbot_control_system.distance0,jellyfishbot_control_system.distance))
        #time.sleep(0.1)
        # info = Vector3()
        # info.x = xd
        # info.y = yd
        # info.z = u_d
        # run_info.publish(info)
    elif MODE == 3:
        #print('555')
        # waypoints
        global first_time2
        global following_point
        global point1
        # path_points[0,0] = x
        # path_points[1,0] = y
        first_time = True

        ####
        delta = 1
        #if  first_time2:
        if point1 < Npoints:
            ax = path_points[0,point1]
            ay = path_points[1,point1]
            yf = path_points[1,point1+1]
            xf = path_points[0,point1+1] 
        else:
            jellyfishbot_control_system.stop = True
            tau_L,tau_R,tau_M = 0,0,0
            if jellyfishbot_control_system.enable_print:
                print("I finished the paaath")
  
        #jellyfishbot_control_system.Yp = Yp 
        jellyfishbot_control_system.ax = ax
        jellyfishbot_control_system.ay = ay
        jellyfishbot_control_system.yf = yf
        jellyfishbot_control_system.xf = xf 
        jellyfishbot_control_system.delta = delta

        
            

        projected_point = jellyfishbot_control_system.projected_point3()

        yp = projected_point[0] #+ ay
        xp = projected_point[1]


        jellyfishbot_control_system.xp = xp
        jellyfishbot_control_system.yp = yp

        yd,xd = jellyfishbot_control_system.virtual_target2()
     
     
        if jellyfishbot_control_system.enable_print:
            print("ax,ay = {},{}".format(jellyfishbot_control_system.ax,jellyfishbot_control_system.ay))
            print("xp,yp = {},{}".format(xp,yp))
            print("xd,yd = {},{}".format(xd,yd))
            print("xf,yf = {},{}".format(jellyfishbot_control_system.xf,jellyfishbot_control_system.yf))
        
            
        counter = counter +1
        
        u_d= 0.5
        # cross_track_error = sqrt((xp - x)**2+(yp - y)**2)
        # #print("Cross track error: {}".format(cross_track_error))
        # if jellyfishbot_control_system.distance < 1 or cross_track_error > 0.5 or counter < 100 :
        #     u_d = 0.1
        # elif 100 <= counter < 150:
        #     u_d = 0.3
        #print("counter: {}".format(counter))
  
  
        if jellyfishbot_control_system.distance < 0.5  and counter > 100 or jellyfishbot_control_system.getting_away:
            point1 = point1+1
            counter = 0
        
        v_d= 0
        jellyfishbot_control_system.u_d_yaw = 5 # desired surge velocity for periods of yaw motion


        threshold = 0.5
        jellyfishbot_control_system.xd = xd 
        jellyfishbot_control_system.yd = yd
        jellyfishbot_control_system.u_d = u_d
        jellyfishbot_control_system.v_d = v_d
        if not jellyfishbot_control_system.check_arrival(threshold) or 1 <2:
            
            #tau_L,tau_R,tau_M = jellyfishbot_control_system.obtain__thruster_commands_LOS_V4(xd,yd,u_d,v_d)
            tau_L,tau_R,tau_M = jellyfishbot_control_system.obtain__thruster_commands_LOS_Virtual_target(u_d)
            
            #tau_R = min(tau_R,35)
            #tau_L = min(tau_L,35)
            
            pub_thrust_l.publish(tau_R)     
            pub_thrust_r.publish(tau_L)     
            pub_thrust_t.publish(tau_M)
            if jellyfishbot_control_system.enable_print:
                print("thrusters: {},{},{},    {}, {},{},   {}  {}    {}".format(tau_L,tau_R,tau_M,round(jellyfishbot_control_system.psi_d,3),round(jellyfishbot_control_system.psi,3),round(jellyfishbot_control_system.e_psi,3),jellyfishbot_control_system.movement,round(jellyfishbot_control_system.distance,2),round(jellyfishbot_control_system.u,2)))
                
        else:
            following_point = following_point +1
            #jellyfishbot_control_system.stop = True
            #tau_L,tau_R,tau_M = 0,0,0
            if jellyfishbot_control_system.enable_print:
                 print("I reach the point {},{}, haaay".format(jellyfishbot_control_system.xf,jellyfishbot_control_system.yf))
 
        # jellyfishbot_control_system.distance0 = distance0
        # print("distance0,distance: {},{}".format(jellyfishbot_control_system.distance0,jellyfishbot_control_system.distance))
        #time.sleep(0.1)
        # info = Vector3()
        # info.x = xd
        # info.y = yd
        # info.z = u_d
        # run_info.publish(info)

def listener():
    
    # rospy.init_node('listener', anonymous=False)
    # rate = rospy.Rate(10)

    

    rospy.Subscriber("/robot_twist_bff", Twist, velocity_callback)
    rospy.Subscriber("/robot_pose", Pose, pose_callback)
    # rate.sleep()
    
    rospy.spin()

if __name__ == '__main__':
    rospy.init_node('listener', anonymous=False)
    if command_thrusters ==1:
        pub_thrust_l = rospy.Publisher('/thrust_l', Float32, queue_size=10)
        pub_thrust_r = rospy.Publisher('/thrust_r', Float32, queue_size=10)
        pub_thrust_t = rospy.Publisher('/thrust_t', Float32, queue_size=10)
        # rospy.Timer(rospy.Duration(1.0/10.0), pub_thrust_l)
        # rospy.Timer(rospy.Duration(1.0/10.0), pub_thrust_r)
        # rospy.Timer(rospy.Duration(1.0/10.0), pub_thrust_t)
    else:
        pub_thrust_l = rospy.Publisher('/thrust_l_off', Float32, queue_size=10)
        pub_thrust_r = rospy.Publisher('/thrust_r_off', Float32, queue_size=10)
        pub_thrust_t = rospy.Publisher('/thrust_t_off', Float32, queue_size=10) 
    run_info = rospy.Publisher('/run_info', Vector3, queue_size=10)
    
    listener()
